File: Scientific_Visualization/codes/writeNetCDF.f90
program writeNetCDF

  ! this program writes data in NetCDF

  use netcdf
  implicit none
  integer, parameter :: nx = 100, ny = nx, nz = nx
  character*90, parameter :: filename = 'volume.nc'
  integer :: ncid, varid, dimids(3), x_dimid, y_dimid, z_dimid, i, j, k
  real*4:: x, y, z, r
  real*4, parameter :: r0 = 0.4
  real*4, dimension(:,:,:), allocatable :: data

  allocate(data(nx,ny,nz))

  do i = 1, nx
     x = (float(i)-0.5)/float(nx)
     do j = 1, ny
        y = (float(j)-0.5)/float(ny)
        do k = 1, nz
           z = (float(k)-0.5)/float(nz)
           r = sqrt((x-0.5)**2+(y-0.5)**2)
           data(i,j,k) = exp(-((z-0.5)**2+(r-r0)**2)**0.5)
        enddo
     enddo
  enddo

  call check(nf90_create(trim(filename), NF90_CLOBBER, ncid))
  call check(nf90_def_dim(ncid, "z", nz, z_dimid))
  call check(nf90_def_dim(ncid, "y", ny, y_dimid))
  call check(nf90_def_dim(ncid, "x", nx, x_dimid))
  dimids =  (/ x_dimid, y_dimid, z_dimid /)
  call check(nf90_def_var(ncid, "density", NF90_FLOAT, dimids, varid))
  call check(nf90_enddef(ncid))
  call check(nf90_put_var(ncid, varid, data))
  call check(nf90_close(ncid))

contains

  subroutine check(status)

    integer, intent (in) :: status
    
    if (status.ne.nf90_noerr) then 
      print *, trim(nf90_strerror(status))
      stop 2
    end if

  end subroutine check

end program writeNetCDF

File: Scientific_Visualization/codes/writeNetCDF.cpp
#include <iostream>
#include <netcdf>
#include <fstream>
#include <math.h>
using namespace std;
using namespace netCDF;
using namespace netCDF::exceptions;

/* this program writes data in NetCDF */

#define nx 100
#define ny nx
#define nz nx
#define NC_ERR 2

float data[nx][ny][nz]; // neeed contiguous storage for NetCDF, so no dynamic allocation; allocate in global heap

int main() {
  int ncid, x_dimid, y_dimid, z_dimid;
  int dimids[3], retval, i, j, k;
  float x, y, z, r, r0 = 0.4;

  for (i = 0; i < nx; i++) {
    if (nx > 100 && i%100 == 0) cout << i << endl;
    x = ((float)i+0.5)/(float)nx;
    for (j = 0; j < ny; j++) {
      y = ((float)j+0.5)/(float)ny;
      for (k = 0; k < nz; k++) {
	z = ((float)k+0.5)/(float)nz;
	r = sqrt(pow(x-0.5,2)+pow(y-0.5,2));
	data[i][j][k] = exp(-sqrt(pow(z-0.5,2)+pow(r-r0,2)));
      }
    }
  }

  try {
    NcFile fileHandle("volume.nc", NcFile::replace);
    NcDim xDim = fileHandle.addDim("x", nx);
    NcDim yDim = fileHandle.addDim("y", ny);
    NcDim zDim = fileHandle.addDim("z", nz);
    vector<NcDim> dims;
    dims.push_back(xDim);
    dims.push_back(yDim);
    dims.push_back(zDim);
    NcVar varID = fileHandle.addVar("density", ncFloat, dims);
    varID.putVar(data);
    return 0;
  }
  catch(NcException& e) {
    e.what();
    return NC_ERR;
  }
}
